### ISSUE TITLE
ERROR Unexpected exception: OSError(17, 'File exists')

### ISSUE THREAD

**Issue by mzeidhassan** [ModernMT/MMT#285]
Hello MMT team,

First, congrats on the neural decoder. This is a great milestone indeed. 

I thought I should give it a try although I don't have  a GPU card yet. I believe MMT should be able to create a neural engine using CPU, but it will take way longer to train it, right?

So, I installed the latest version of MMT, but initially when I used:

`./mmt create en ar examples/data/train --neural `

I got the following error:

> ERROR Illegal Argument: you project does not include EN-AR data.

then, I removed all sample training data and uploaded new versions of the text files 'english.en' and 'arabic.ar' in UTF-8 format, but I keep getting this error:

"ERROR Unexpected exception: OSError(17, 'File exists')"

any idea?

Thanks in advance for your support!
Mohamed

**Comment by AndRossi** [comment]
Hello @mzeidHassan,

Thanks for trying our new Neural MMT.

As you thought, you can train a neural engine using your CPU instead of your GPU, but it will take extremely longer: even on a tiny dataset, you will probably need to wait several hours before it completes. 
This happens because neural networks involve lots of matrix calculations, and that is something that CPUs are really slow at doing, while GPUs can do that lightning-fast.


As for the  `ERROR Illegal Argument: you project does not include EN-AR data.`: by default our `examples/data/train` folder only contains a tiny en-it corpus. We include as an example of how a correct corpus folder looks like. So if you want to create an en-ar engine on it, your engine will complain because it will not find any suitable en-ar data.  
Instead of `examples/data/train` just use the path to the folder where your en-ar corpora are kept, like: `./mmt create en ar /path/tp/your/en-ar/corpora`  
Of course copying your data to the `examples/data/train` folder will work as well, but it is less efficient :) 

As for the `OSError(17, 'File exists')` error, could you please run again the `./mmt create` command adding `--debug` option too, and send me the resulting `<your_mmt_home>/runtime/<your_engine_name>/logs/training.log` file? 

Thanks for your patience and support!

Andrea

**Comment by mzeidhassan** [comment]
Hi @AndRossi ,

Thanks for your reply and explanation. I appreciate it.
For some reason, when I added 'sudo' to the command, it worked.

It's worth-mentioning though that I got this message at step 7 "Neural decoder training..."

`INFO: (7 of 7) Neural decoder training...                             /home/                     ModernMT/build/lib/pynmt/nmmt/NMTEngineTrainer.py:177: UserWarning: self and other no                     t broadcastable, but have the same number of elements.  Falling back to deprecated po                     intwise behavior.
  num_correct_t = pred_t.data.eq(targ_t.data).masked_select(targ_t.ne(Constants.PAD).                     data).sum()
/home/ModernMT/build/lib/pynmt/nmmt/NMTEngineTrainer.py:177: UserWarning: se                     lf and mask not broadcastable, but have the same number of elements.  Falling back to                      deprecated pointwise behavior.
  num_correct_t = pred_t.data.eq(targ_t.data).masked_select(targ_t.ne(Constants.PAD).                     data).sum()
`
but it seems it's still running? What do you think? Does the error message above indicate a problem or maybe a bug?

Thanks,
Mohamed

**Comment by AndRossi** [comment]
Hi @mzeidhassan, 

I'm glad you managed to run the training.
If with `sudo` it went well, the cause was probably an issue with the permissions in one of the mmt folders.

As for the message, that is not a problem. It reports that, at the moment, we are using a deprecated function from pytorch. Since this does not interfere with the training process, we decided to ignore it for the time being :)

Please let me know if the training ends smoothly.

Have a nice day!

Andrea

**Comment by mzeidhassan** [comment]
Thanks @AndRossi for the update. Glad to know that the message can be ignored :)

I will let you know if I have any problems with training.

I have another question about the volume that an NMT engine needs. For SMT, we know that the bigger, the better.  

For NMT, what do you recommend in terms of number of parallel strings to get high quality results?

Thanks in advance for your support!

Kind regards,
Mohamed

**Comment by EtienneAb3d** [comment]
The bigger is not always the better, whatever the tools. It depends on what you are expecting. If, you want to translate in-domain sentences, you need to add texts from this domain. The nearer the better. The lack of quantity will force you to add generic data to gain on unknown formulations and/or unknown in-domain vocabulary, but with certainly a degradation of in-domain translations. I guess this is also true here in spite of the on-the-fly retraining of MMT..
;-)

**Comment by AndRossi** [comment]
Hi @Mzeidhassan,

As @Etienne pointed out, the amount of data is not always the only parameter to take into account.

In most machine translation systems, adding new generic data leads to a degradation of the in-domain translations. 

ModernMT, on the contrary, is a context-aware system, so it does not suffer from this problem.
If you have trained your system with data for a certain domain, adding new generic contents should not affect the quality of translations for that domain, because when translating in-domain content, that domain will be identified as the current context, and it will be taken into account with higher priority.

In our tests we obtained good translation results with corpora containing 100M source words.
The amount of words you want to use, however, does also depend on the quality of your corpora, on your language direction and vocabulary, etc.

**Comment by mzeidhassan** [comment]
Thank you @EtienneAb3d and @AndRossi for your explanation. I appreciate it.
I will follow a different approach from now on based on your advice. Thanks again!

**Comment by EtienneAb3d** [comment]
Dear @AndRossi, just an example...
A trained a neural OpenNMT model with 2M generic sentences + 200K in-domain sentences. In-domain sentences were duplicated to be about 20% of the total volume. This pragmatic trade-off ensure that the model will be as close to the in-domain data as possible, but also will take into account some other kinds of formulations. I got a BLEU of 48 on in-domain validation data (of course, not already in the training data).
A then similarly trained a neural MMT model with 20M generic sentences + 200K in-domain sentences. In this case, in-domain sentences weren't duplicated, hoping that MMT will do the job with its post domain-specific calculation. I got a BLEU of 31 on in-domain validation data.
Comparing the results, MMT translations weren't so bad. They were in fact, quite good. But, they show much more variations in the details of formulations, and in local or global order of words groups. 
This neither means that BLEU is a bad evaluation, nor that MMT didn't improve the results. It only shows that, in spite of the in-domain post retraining of MMT, its translations were less close to the in-domain references... but, in fact, still good, or even better (in a human point of view).
But.. possibly, there is also something to tune in MMT to increase the domain separation made by its post-processing... Is this tuning possible ?
:-)

**Comment by AndRossi** [comment]
Hello @mzeidhassan, 

In the end, did your training run smoothly? :)

Andrea

**Comment by mzeidhassan** [comment]
Thanks @AndRossi for your follow-up. It's still running actually because I am using CPU :-( . 
I am planning to buy a new machine with Nvidia GTX 1080 TI and I hope things will be faster. If you wish, you can close this issue for now and I will keep you posted if something wrong happens. Thanks again and have a wonderful weekend ahead!

Kind regards,
Mohamed

**Comment by AndRossi** [comment]
Hi @mzeidhassan, 

As you suggested, since the `OSError(17, 'File exists')` problem was solved I am going to close this issue now.
If any other problem or error occurs, feel free to reopen this issue or to create new ones :) 

Have a nice day and thanks again for your support!

Andrea
